### PR TITLE
DC-Sync: Fill indexTime and LastModifiedBy

### PR DESCRIPTION
**Pull request by MoriaYechiel**
Make DC-Sync fill indexTime and LastModifiedBy if not exist

### CODE REVIEW

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 1]
```diff
@@ -55,15 +53,32 @@ object InfotonRetriever extends LazyLogging {
...
     RetrieveStateStatus(Settings.initialSingleRetrieveRetryCount, None)
 
-  case class ParsedNquad(path: String, nquad: ByteString, indexTime: Option[Long])
+  case class ExtraData(subject: String, indexTime: Option[Long], lastModifiedBy: Option[String]){
+    def shouldMerge (other: ExtraData): Option[ExtraData] = {
+      other match {
+        case ExtraData("", None, None) => None
+        case ExtraData(newSubject, None, None) if this.subject == "" => Some(this.copy(subject = newSubject))
+        case ExtraData(_, None, None) => None
+        case ExtraData(newSubject, Some(_), None) => Some(this.copy(subject = newSubject, indexTime = other.indexTime))
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: The got subject should never be different from the current one - don't override it. Instead, throw an exception.

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 2]
```diff
@@ -55,15 +53,32 @@ object InfotonRetriever extends LazyLogging {
...
+        case ExtraData(newSubject, None, Some(_)) => Some(this.copy(subject = newSubject, lastModifiedBy = other.lastModifiedBy))
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: same

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 1 above)
Comment (e-orz) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: It will be nicer to use the value got inside the Some (instead of _ and using `other`)

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 3]
```diff
@@ -55,15 +53,32 @@ object InfotonRetriever extends LazyLogging {
...
+      }
+    }
+
+    def getQuads(indexTimeN: Long, modifierN: String) = {
+      indexTime.fold(s"""<$subject> <cmwell://meta/sys#indexTime> "${indexTimeN}"^^<http://www.w3.org/2001/XMLSchema#long> .""")(_ => "") ++
+      lastModifiedBy.fold(s"""<$subject> <cmwell://meta/sys#lastModifiedBy> "$modifierN" .""" )(_ => "")
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: You might want to add a Sting schema (for completeness)

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 4]
```diff
@@ -96,33 +111,25 @@ object InfotonRetriever extends LazyLogging {
...
-              ) { (totals, nquad) =>
+              .fold(RetrieveTotals(HashMap(state._1 map{im => im.meta.path -> (new ByteStringBuilder, ExtraData("", None, None))}: _*), new ByteStringBuilder))
+              { (totals, nquad) =>
                 totals.unParsed ++= nquad
-                val parsed: Option[ParsedNquad] = parseNquad(remoteUri, nquad)
-                parsed.fold {
-                  RetrieveTotals(totals.parsed, totals.unParsed)
-                } { p =>
-                  totals.parsed.get(p.path) match {
+                val parsed: ParsedNquad = parseNquad(remoteUri, nquad)
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: there are lines got from _out that are not related to any infoton. You must return Option or so to "know" about it.

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 5]
```diff
@@ -96,33 +111,25 @@ object InfotonRetriever extends LazyLogging {
...
-                      val parsedTotals = p.indexTime.fold(totals.parsed)(
-                        _ => totals.parsed + (p.path -> (builder, p.indexTime))
-                      )
+                      throw WrongPathGotException(s"Got path ${path} from _out that was not in the uuids request bulk: ${
+                        state._1.map(i => i.meta.uuid.utf8String + ":" + i.meta.path).mkString(",")}")
+                    case Some((builder, curExtraData)) => {
+                      builder ++= (nquad ++ endln)
+                      val parsedTotals = {
+                        val merged = curExtraData.shouldMerge(extraData)
+                        merged.fold(totals.parsed)
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: You don't use the returned value. This statement has side effects only - please use `foreach` instead.

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 6]
```diff
@@ -96,33 +111,25 @@ object InfotonRetriever extends LazyLogging {
...
               )
-              .fold(
-                RetrieveTotals(
-                  state._1
-                    .map(im => (im.meta.path, (new ByteStringBuilder, None)))
-                    .toMap,
-                  new ByteStringBuilder
-                )
-              ) { (totals, nquad) =>
+              .fold(RetrieveTotals(HashMap(state._1 map{im => im.meta.path -> (new ByteStringBuilder, ExtraData("", None, None))}: _*), new ByteStringBuilder))
```
[the remark this answers is not in the document]
Comment (MoriaYechiel) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: Won't compile without

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 7]
```diff
@@ -481,6 +481,7 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
               case JsDefined(JsArray(seq))
                 if seq.length == 1 && seq.head.isInstanceOf[JsString] =>
                 seq.head.as[String]
+              case _ => ""
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: This should be Option (or any other way). Otherwise, it will add an empty modifier for infotons that doesn't have last-modified-by.
